### PR TITLE
adding linprog methods

### PR DESCRIPTION
**Pull request by RenanOD**

### CODE REVIEW

**Comment by RenanOD** [comment]
Currently linprog supports the following methods:
`linprog(c, A, b)`
`linprog(c, A, b, Aeq, beq)`
`linprog(c, A, b, Aeq, beq, lvar, uvar)`
But `lvar` and `uvar` could be passed as keyword arguments, allowing
`linprog(c, A, b, lvar = lvar, uvar = uvar)`.

An initial solution could also be passed as keyword argument.

**Comment by abelsiqueira** [comment]
Yes, please, allow an initial solution and keyword arguments.
You should also have one method like `linprog(c, A, blow, bupp)`.

**Comment by RenanOD** [comment]
I plan on making
`linprog(c, A, bup; blow = [-Inf ...], uvar = [+Inf ...], lvar = [-Inf ...], x0 = zeros)`
`linprog(c, A, bup, Aeq, beq, uvar = [+Inf ...], lvar = [-Inf ...]; blow = [-Inf ...], x0 = zeros)`

This should allow all variations and is consistent with the usual syntax.

File: src/linprog.jl [hunk 1]
```diff
@@ -3,46 +3,45 @@ export linprog
...
-  ucon = [b; beq]
-  nlp = ADNLPModel(f, zeros(length(c)), c = con, lcon = lcon, ucon = ucon)
-  output = ipopt(nlp, print_level=0)
-  return output
-end
+Minimize cᵀ x subject to A x ≤ bup and Aeq x = beq.
 
-"""
-    linprog(c, A, b, Aeq, beq, lvar, uvar)
+    linprog(c, A, bup, Aeq, beq, uvar, lvar)
```
Comment (abelsiqueira) [inline src/linprog.jl]: Order: `lvar, uvar`

**Comment by abelsiqueira** [comment]
I didn't think you would remove the positional variants. I think you can have one fully made of keyword arguments (except for the objective): `linprog(c; ...)`, and the create other methods with positional non-optional arguments. They could call the general one, e.g.:
- `linprog(c, A, b) = linprog(c, Aeq=A, beq=b)`
- `linprog(c, A, blow, bupp) = linprog(c, Aineq=A, blow=blow, bupp=bupp)`

**Comment by RenanOD** [comment]
> I didn't think you would remove the positional variants. I think you can have one fully made of keyword arguments (except for the objective): `linprog(c; ...)`, and the create other methods with positional non-optional arguments. They could call the general one, e.g.:
> 
> * `linprog(c, A, b) = linprog(c, Aeq=A, beq=b)`
> * `linprog(c, A, blow, bupp) = linprog(c, Aineq=A, blow=blow, bupp=bupp)`

I like this idea, will be doing that.

**Comment by RenanOD** [comment]
Should be ready for review.

File: src/linprog.jl [hunk 2]
```diff
@@ -1,16 +1,67 @@
...
+
+Minimize cᵀ x subject to any combination of the optional constraints
+blow ≤ A x ≤ bup, Aeq x = beq and lvar ≤ x ≤ uvar.
+Initial solution is optional and set to zeros by default.
+One sided inequalities like A x ≤ bup are equivalent to [-∞,…,-∞] ≤ A x ≤ bup.
 
-Minimize the linear problem cᵀ x with constraints A x ≤ b.
 """
-function linprog(c::AbstractVector, A::MatrixOrOperator, b::AbstractVector)
+function linprog(c::AbstractVector; x0::AbstractVector = zeros(length(c)), A::MatrixOrOperator = zeros(0,0),
```
Comment (abelsiqueira) [inline src/linprog.jl]: A few things:
- Change `up` to `upp`;
- Change initial values: `A = zeros(0, length(c))`, `Aeq = zeros(0, length(c))`;
- Add `@assert` checks for size consistency;
- `ADNLPModel` defines the number of constraints according to the size of `lcon` and `ucon`, so only line 33 is necessary.

Comment (abelsiqueira) [reply src/linprog.jl]: Also, change `A` to `Aineq`.

**Comment by abelsiqueira** [comment]
Thanks for the PR. You're still missing a few methods.

**Comment by RenanOD** [comment]
> Thanks for the PR. You're still missing a few methods.

Which ones should I do?

**Comment by abelsiqueira** [comment]
- `linprog(c, A, b)` which means `Aineq=A` and `bupp = b`;
- `linprog(c, A, b, Aeq, beq)`;
- `linprog(c, A, b, lvar, uvar)`;
- `linprog(c, A, b, Aeq, beq, lvar, uvar)`;

**Comment by RenanOD** [comment]
Ready for review again, I removed some of the methods.
Now there are

* `linprog(c; x0, Aineq, blow, bupp, Aeq, beq, lvar, uvar) `;
> * `linprog(c, A, b)` which means `Aineq=A` and `bupp = b`;
> * `linprog(c, A, b, Aeq, beq)`;
> * `linprog(c, A, b, lvar, uvar)`;
> * `linprog(c, A, b, Aeq, beq, lvar, uvar)`;
* `linprog(c, A, blow, bupp)`;

File: src/linprog.jl [hunk 3]
```diff
@@ -1,16 +1,87 @@
...
+  return linprog(c, Aineq = A, bupp = b, x0 = x0)
+end
+
+"""
+    linprog(c, A, b, Aeq, beq; x0)
+
+Minimize cᵀ x subject to A x ≤ b and Aeq x = beq.
+"""
+function linprog(c::AbstractVector, A::MatrixOrOperator, b::AbstractVector, Aeq::MatrixOrOperator, beq::AbstractVector;
+         x0::AbstractVector = zeros(length(c)))
```
Comment (abelsiqueira) [inline src/linprog.jl]: alignment

File: src/linprog.jl [hunk 4]
```diff
@@ -1,16 +1,87 @@
...
+  return linprog(c, Aineq = A, bupp = b, Aeq = Aeq, beq = beq, x0 = x0)
+end
+
+"""
+    linprog(c, A, b, lvar, uvar; x0)
+
+Minimize cᵀ x subject to A x ≤ b and lvar ≤ x ≤ uvar.
+"""
+function linprog(c::AbstractVector, A::MatrixOrOperator, b::AbstractVector,
+         lvar::AbstractVector, uvar::AbstractVector; x0::AbstractVector = zeros(length(c)))
```
Comment (abelsiqueira) [inline src/linprog.jl]: alignment

File: src/linprog.jl [hunk 5]
```diff
@@ -1,16 +1,87 @@
...
+  return linprog(c, Aineq = A, bupp = b, lvar = lvar, uvar = uvar, x0 = x0)
+end
+
+"""
+    linprog(c, A, b, Aeq, beq, lvar, uvar; x0)
+
+Minimize cᵀ x subject to A x ≤ b, Aeq x = beq and lvar ≤ x ≤ uvar.
+"""
+function linprog(c::AbstractVector, A::MatrixOrOperator, b::AbstractVector, Aeq::MatrixOrOperator,
+         beq::AbstractVector, lvar::AbstractVector, uvar::AbstractVector; x0::AbstractVector = zeros(length(c)))
```
Comment (abelsiqueira) [inline src/linprog.jl]: alignment

File: src/linprog.jl [hunk 6]
```diff
@@ -1,16 +1,87 @@
...
+  return linprog(c, Aineq = A, bupp = b, Aeq = Aeq, beq = beq, lvar = lvar, uvar = uvar, x0 = x0)
+end
+
+"""
+    linprog(c, Aineq, blow, bupp; x0)
+
+Minimize cᵀ x subject to blow ≤ Aineq x ≤ bupp.
+"""
+function linprog(c::AbstractVector, Aineq::MatrixOrOperator, blow::AbstractVector, bupp::AbstractVector;
+         x0::AbstractVector = zeros(length(c)))
```
Comment (abelsiqueira) [inline src/linprog.jl]: alignment

File: test/linprog.jl [hunk 7]
```diff
@@ -1,18 +1,52 @@
 function linprog_tests()
   @testset "linprog" begin
+
+    # testing with Aineq
```
Comment (abelsiqueira) [inline test/linprog.jl]: Remove this comment and the additional line

File: test/linprog.jl [hunk 8]
```diff
@@ -1,18 +1,52 @@
...
-    b = [2.; 1; 2; 1; -1; 2]
+    Aineq = [1. 1; 1 1/4; 1 -1; -1/4 -1; -1 -1; -1 1]
+    opAineq = LinearOperator(Aineq)
+    bupp = [2.; 1; 2; 1; -1; 2]
+    blow = fill(-Inf, length(bupp))
     sol = [2/3; 4/3]
-    output = linprog(c, A, b)
+
+    output = linprog(c, Aineq, bupp)
+    @test output.dual_feas < 1e-6
```
Comment (abelsiqueira) [inline test/linprog.jl]: Also test `output.primas_feas < 1e-6`.

**Comment by RenanOD** [comment]
> Thanks again. I've a few more comments, if you don't mind. Probably the last of it.

No problem. It's ready for review again.

File: src/linprog.jl [hunk 9]
```diff
@@ -62,7 +51,7 @@ end
 Minimize cᵀ x subject to A x ≤ b and lvar ≤ x ≤ uvar.
 """
 function linprog(c::AbstractVector, A::MatrixOrOperator, b::AbstractVector,
-         lvar::AbstractVector, uvar::AbstractVector; x0::AbstractVector = zeros(length(c)))
+		 lvar::AbstractVector, uvar::AbstractVector; x0::AbstractVector = zeros(length(c)))
```
Comment (abelsiqueira) [inline src/linprog.jl]: I'm not sure what happened here and in other places.

Comment (abelsiqueira) [reply src/linprog.jl]: Apparently, you changed to TAB and that broke it? Use spaces, you can configure your editor to use spaces in TABs.

Comment (RenanOD) [reply src/linprog.jl]: Sorry, my editor does that sometimes. It was fixed by the last commit.

**Comment by abelsiqueira** [comment]
Thank you!

**Pull request merged**